packages/appium/lib/extension/driver-config.js

```diff
@@ -0,0 +1,215 @@
+// @ts-check
+
+import _ from 'lodash';
+import { DRIVER_TYPE } from '../constants';
+import log from '../logger';
+import { ExtensionConfig } from './extension-config';
+
+/**
+ * @extends {ExtensionConfig<DriverType>}
+ */
+export class DriverConfig extends ExtensionConfig {
+  /**
+   * A mapping of `APPIUM_HOME` values to {@link DriverConfig} instances.
+   * Each `APPIUM_HOME` should only have one associated `DriverConfig` instance.
+   * @type {Record<string,DriverConfig>}
+   * @private
+   */
+  static _instances = {};
+
+  /**
+   * Call {@link DriverConfig.getInstance} instead.
+   * @private
+   * @param {import('./manifest').Manifest} io - IO object
+   * @param {DriverConfigOptions} [opts]
+   */
+  constructor (io, {logFn, extData} = {}) {
+    super(DRIVER_TYPE, io, logFn);
+    /** @type {Set<string>} */
+    this.knownAutomationNames = new Set();
+
+    if (extData) {
+      this.validate(extData);
+    }
+  }
+
+  /**
+   * Checks extensions for problems
+   * @param {ExtRecord<DriverType>} exts
+   */
+  validate (exts) {
+    this.knownAutomationNames.clear();
+    return super.validate(exts);
+  }
+
+  /**
+   * Creates a new DriverConfig
+   *
+   * Warning: overwrites any existing `DriverConfig` for the given `appiumHome` prop of the `io` parameter.
+   * @param {import('./manifest').Manifest} io
+   * @param {DriverConfigOptions} [opts]
+   * @returns {DriverConfig}
+   */
+  static create (io, {extData, logFn} = {}) {
+    const instance = new DriverConfig(io, {logFn, extData});
+    DriverConfig._instances[io.appiumHome] = instance;
+    return instance;
+  }
+
+  /**
+   * Gets an existing instance of {@link DriverConfig} based value of `io.appiumHome`
+   * @param {import('./manifest').Manifest} io - IO object
+   * @returns {DriverConfig}
```

Comment:
could the result be undefined?

Reply:
No.  If one does not exist, it will create a new one.

Reply:
but this particular call does not check if `_instances` mapping contains `io.appiumHome`, does it?

Reply:
As you mentioned here and below, these were confusing.  I rewrote them (pushing in a few), so:

- `DriverConfig.create(manifest, opts)` only creates an instance, and throws if one is already associated with a `Manifest`.
- `DriverConfig.getInstance(manifest)` returns an existing instance if it exists, or `undefined` if it does not.

(same for `PluginConfig`)

The mapping is now a `WeakMap` based on the `Manifest` instance instead of `appiumHome`.

Reply:
`loadExtensions()` now tries `getInstance()` before calling `create()`.